keras/engine/topology.py

```diff
@@ -2746,102 +2746,120 @@ def preprocess_weights_for_loading(layer, weights,
-                                                   weights[4],
-                                                   weights[7]], axis=-1)
-                bias = np.concatenate([weights[2],
-                                       weights[5],
-                                       weights[8]], axis=-1)
-                weights = [kernel, recurrent_kernel, bias]
-
-        if layer.__class__.__name__ == 'LSTM':
-            if len(weights) == 12:
-                # old: i, c, f, o
-                # new: i, f, c, o
-                kernel = np.concatenate([weights[0],
-                                         weights[6],
-                                         weights[3],
-                                         weights[9]], axis=-1)
-                recurrent_kernel = np.concatenate([weights[1],
-                                                   weights[7],
-                                                   weights[4],
-                                                   weights[10]], axis=-1)
-                bias = np.concatenate([weights[2],
-                                       weights[8],
-                                       weights[5],
-                                       weights[11]], axis=-1)
-                weights = [kernel, recurrent_kernel, bias]
-
-        if layer.__class__.__name__ == 'ConvLSTM2D':
-            if len(weights) == 12:
-                kernel = np.concatenate([weights[0],
-                                         weights[6],
-                                         weights[3],
-                                         weights[9]], axis=-1)
-                recurrent_kernel = np.concatenate([weights[1],
-                                                   weights[7],
-                                                   weights[4],
-                                                   weights[10]], axis=-1)
-                bias = np.concatenate([weights[2],
-                                       weights[8],
-                                       weights[5],
-                                       weights[11]], axis=-1)
+
+    if layer.__class__.__name__ == 'Bidirectional':
+        num_weights_per_layer = len(weights) // 2
+
+        forward_weights = preprocess_weights_for_loading(layer.forward_layer,
+                                                         weights[:num_weights_per_layer],
+                                                         original_keras_version,
+                                                         original_backend)
+        backward_weights = preprocess_weights_for_loading(layer.backward_layer,
+                                                          weights[num_weights_per_layer:],
+                                                          original_keras_version,
+                                                          original_backend)
+        weights = forward_weights + backward_weights
+    elif layer.__class__.__name__ == 'TimeDistributed':
```

Comment:
I am not sure I follow, I currently have it as:

`if TimeDistributed:
    #recurse
elif Bidirectional:
    #recurse
else:
    #original code` 

Would this be cleaner as a new PR? Because I think a lot of the changes are essentially reverting my original suggestion.